### PR TITLE
[WIP] Add Ascend NPU support for DSpark training

### PR DESCRIPTION
**Pull request by hazelduan**
This draft PR adds Ascend NPU support for the Gemma4 DSpark training and evaluation path.

It introduces a small accelerator abstraction so the code can select CUDA or NPU at runtime, using HCCL for NPU distributed training. For Gemma4 DSpark attention, the training path uses PyTorch SDPA with a dense boolean mask fallback, because the original FlexAttention BlockMask path is not available in the current torch_npu stack. The dense mask preserves DSpark’s draft/block masking semantics, but it is a correctness baseline rather than the final block-sparse performance path.

Validation performed:
Prepared a Gemma4 target cache on Ascend NPU with 1024 training samples.
Ran DSpark Gemma4 training for 100 steps on 8 NPU devices.
Confirmed checkpoint save at step 100.
Ran speculative evaluation smoke test with the trained draft checkpoint.

100-step training trend:
train/loss decreased from 3.3606 at step 10 to 2.3759 at step 100.
train/tau_probabilistic, used here as the acceptance/reward proxy, increased from 1.0802 to 1.2766.
Mean accept rate increased from about 6.9% to 13.1%.
accept_rate@0 increased from about 7.0% to 24.3%.

This indicates the Gemma4 DSpark draft path is functional on Ascend NPU and the training signal moves in the expected direction.

### CODE REVIEW

File: deepspec/eval/dspark/confidence_head.py [hunk 1]
```diff
@@ -43,20 +43,20 @@ def __init__(
         self.num_fine_bins = int(num_fine_bins)
         self.coarse_count = torch.zeros(
             (self.block_size, self.num_coarse_bins),
-            dtype=torch.float64,
+            dtype=torch.float32,
```
Comment (codemayq) [inline deepspec/eval/dspark/confidence_head.py]: 原则上不能修改这种硬精度逻辑，作者本身会有其考虑。如果要兼容，把 dtype 抽象成一个 变量，GPU 还是保留为float64，NPU 为fp32。

File: deepspec/eval/dspark/confidence_head.py [hunk 2]
```diff
@@ -539,9 +539,12 @@ def log_tensorboard(self) -> None:
         writer.close()
 
     def build_table(self) -> str:
-        from prettytable import PrettyTable
+        try:
+            from prettytable import PrettyTable
```
Comment (codemayq) [inline deepspec/eval/dspark/confidence_head.py]: 这个也有依赖问题么

File: deepspec/eval/dspark/confidence_head.py [hunk 3]
```diff
@@ -568,36 +573,42 @@ def build_table(self) -> str:
...
-                        format_float(per_position[pos]["auc"])
-                        if pos in per_position
-                        and float(per_position[pos]["total_weight"]) > 0.0
-                        else "-"
-                    )
-                    for pos in range(max_position_count)
-                ]
-            )
-        return table.get_string()
+            table_row = [
```
Comment (codemayq) [inline deepspec/eval/dspark/confidence_head.py]: 这里check下 修改的必要性

File: deepspec/eval/dspark/evaluator.py [hunk 4]
```diff
@@ -182,11 +184,18 @@ def evaluate(self) -> None:
         for dataset_name, max_samples in self.tasks:
             if self.confidence_head_recorder is not None:
                 self.confidence_head_recorder.start()
+            dist.barrier()
+            start_time = time.perf_counter()
             responses = self.run_dataset(
                 dataset_name=dataset_name,
                 max_samples=max_samples,
             )
+            dist.barrier()
```
Comment (codemayq) [inline deepspec/eval/dspark/evaluator.py]: 整块需要check下修改的必要性

File: deepspec/modeling/dspark/gemma4/config.py [hunk 5]
```diff
@@ -3,7 +3,7 @@
 from deepspec.modeling.dspark.common import validate_target_layer_ids
 
 
-TRAIN_ATTN_IMPLEMENTATION = "flex_attention"
+TRAIN_ATTN_IMPLEMENTATION = "sdpa"
```
Comment (codemayq) [inline deepspec/modeling/dspark/gemma4/config.py]: 不能强制修改原逻辑，如果短期支持不了，需要改为兼容逻辑

File: eval.py [hunk 6]
```diff
@@ -26,6 +16,23 @@
     ("alpaca", 500),
     ("arena-hard-v2", 500),
 ]
+TASK_DEFAULTS = dict(TASKS)
+
+
```
Comment (codemayq) [inline eval.py:29]: 这个保留原逻辑即可，我们作为硬件适配方，不需要去优化这些逻辑，如果要优化，也是另外提PR。

File: deepspec/modeling/dspark/gemma4/config.py (hunk 5 above)
Comment (Seas0) [reply deepspec/modeling/dspark/gemma4/config.py]: 目前NPU是没有FlexAttention支持的，所以应该是要做修改

File: deepspec/modeling/dspark/common.py [hunk 7]
```diff
@@ -82,7 +82,30 @@ def create_dspark_attention_mask(
...
+        is_context = kv_idx < seq_len
+        mask_context = is_context & (kv_idx < anchor_pos)
+        is_draft = kv_idx >= seq_len
+        kv_block_ids = (kv_idx - seq_len) // block_size
+        mask_draft = is_draft & (q_block_ids == kv_block_ids)
+        is_valid_block = block_keep_mask.gather(
+            1,
+            q_block_ids.squeeze(-1),
+        ).unsqueeze(-1)
+        return ((mask_context | mask_draft) & is_valid_block).unsqueeze(1)
```
Comment (huangxinV587) [inline deepspec/modeling/dspark/common.py]: 当某 block 的 block_keep_mask=False 时,is_valid_block(103–106 行)把整条 query 行置 False。SDPA 对全 -inf 行做 softmax 得 NaN(flex_attention 对全屏蔽行返回 0,行为不一致),NaN 经乘法掩码 loss(loss_per_token * weights,NaN×0=NaN)传导成 NaN loss。长序列+小 num_anchors 时不触发;一旦 batch 含有效 anchor 候选数 < num_anchors 的短样本(被 keep_mask=False 填充)就会 NaN。建议在 dense 分支兜底:对全屏蔽行 nan_to_num,或保留对角自身位置(eval_mask 后续会丢弃)。

File: deepspec/eval/base_evaluator.py [hunk 8]
```diff
@@ -154,14 +163,29 @@ def build_results_table(
...
+        else:
+            if not locals().get("fallback_rows"):
+                fallback_rows = []
+            fallback_rows.append(row)
+    if table is not None:
+        return table.get_string()
+    lines = []
+    if header:
+        lines.append("\t".join(str(value) for value in field_names))
+    lines.extend("\t".join(str(value) for value in row) for row in fallback_rows)
```
Comment (huangxinV587) [inline deepspec/eval/base_evaluator.py]: 若 rows 为空,循环不执行,fallback_rows 从未定义。建议本处也在循环前初始化 fallback_rows = []

File: eval.py [hunk 9]
```diff
@@ -42,16 +49,32 @@ def parse_args():
...
+        "--num-samples",
+        type=int,
+        default=None,
+        help="Override the sample count for every selected task.",
+    )
     args = parser.parse_args()
-    args.tasks = list(TASKS)
+    task_names = args.task if args.task else [name for name, _ in TASKS]
+    args.tasks = [
+        (name, args.num_samples if args.num_samples is not None else TASK_DEFAULTS[name])
```
Comment (huangxinV587) [inline eval.py]: 无 --num-samples 会 KeyError

File: deepspec/eval/dspark/confidence_head.py (hunk 2 above)
Comment (hazelduan) [reply deepspec/eval/dspark/confidence_head.py]: fixed in later commit

File: deepspec/modeling/dspark/gemma4/config.py (hunk 5 above)
Comment (hazelduan) [reply deepspec/modeling/dspark/gemma4/config.py]: 已修改为兼容版本

File: deepspec/eval/dspark/confidence_head.py (hunk 1 above)
Comment (hazelduan) [reply deepspec/eval/dspark/confidence_head.py]: Already compatible

File: deepspec/eval/dspark/confidence_head.py (hunk 3 above)
Comment (hazelduan) [reply deepspec/eval/dspark/confidence_head.py]: reverted
